src/modules/dataset.ts

```diff
@@ -19,7 +19,7 @@ function updateDataset(oldVnode: VNode, vnode: VNode): void {
 
   for (key in oldDataset) {
     if (!dataset[key]) {
-      if (d) {
+      if (d && d[key]) {
```

Comment:
this is not enough. this will fail if `elm.dataset.x = 0` or any other falsy value. you will have to change this to `if (d && key in d) {` to only delete it if the key exists.